src/app/clusters/network-commissioning/network-commissioning.cpp

```diff
@@ -18,273 +18,375 @@
+        ctxt.SetCommandHandled();
+        return;
+    }
 
-namespace {
-// The internal network info containing credentials. Need to find some better place to save these info.
-NetworkInfo sNetworks[kMaxNetworks];
-} // namespace
+    // We only use mAsyncCommandHandle to determine whether we are processing another command, so it is OK to left some values
+    // dirty.
 
-void OnAddOrUpdateThreadNetworkCommandCallbackInternal(app::CommandHandler * apCommandHandler,
-                                                       const app::ConcreteCommandPath & commandPath, ByteSpan operationalDataset,
-                                                       uint64_t breadcrumb, uint32_t timeoutMs)
-{
-    Commands::NetworkConfigResponse::Type response;
-#if CHIP_DEVICE_CONFIG_ENABLE_THREAD
-    NetworkCommissioningStatus err = NetworkCommissioningStatus::kBoundsExceeded;
+    mPath = ctxt.mRequestPath;
 
-    for (size_t i = 0; i < kMaxNetworks; i++)
+    switch (ctxt.mRequestPath.mCommandId)
     {
-        if (sNetworks[i].mNetworkType == NetworkType::kUndefined)
-        {
-            Thread::OperationalDataset & dataset = sNetworks[i].mData.mThread;
-            CHIP_ERROR error                     = dataset.Init(operationalDataset);
+    case Commands::ScanNetworks::Id:
+        VerifyOrReturn((mType & kFeatureMapWireless) != 0);
+        HandleCommand<Commands::ScanNetworks::DecodableType>(
+            ctxt, [this](HandlerContext & ctx, const auto & req) { HandleScanNetworks(ctx, req); });
+        return;
+
+    case Commands::AddOrUpdateWiFiNetwork::Id:
+        VerifyOrReturn(mType == kFeatureMapWiFi);
+        HandleCommand<Commands::AddOrUpdateWiFiNetwork::DecodableType>(
+            ctxt, [this](HandlerContext & ctx, const auto & req) { HandleAddOrUpdateWiFiNetwork(ctx, req); });
+        return;
+
+    case Commands::AddOrUpdateThreadNetwork::Id:
+        VerifyOrReturn(mType == kFeatureMapThread);
+        HandleCommand<Commands::AddOrUpdateThreadNetwork::DecodableType>(
+            ctxt, [this](HandlerContext & ctx, const auto & req) { HandleAddOrUpdateThreadNetwork(ctx, req); });
+        return;
+
+    case Commands::RemoveNetwork::Id:
+        VerifyOrReturn((mType & kFeatureMapWireless) != 0);
+        HandleCommand<Commands::RemoveNetwork::DecodableType>(
+            ctxt, [this](HandlerContext & ctx, const auto & req) { HandleRemoveNetwork(ctx, req); });
+        return;
+
+    case Commands::ConnectNetwork::Id:
+        VerifyOrReturn((mType & kFeatureMapWireless) != 0);
+        HandleCommand<Commands::ConnectNetwork::DecodableType>(
+            ctxt, [this](HandlerContext & ctx, const auto & req) { HandleConnectNetwork(ctx, req); });
+        return;
+
+    case Commands::ReorderNetwork::Id:
+        VerifyOrReturn((mType & kFeatureMapWireless) != 0);
+        HandleCommand<Commands::ReorderNetwork::DecodableType>(
+            ctxt, [this](HandlerContext & ctx, const auto & req) { HandleReorderNetwork(ctx, req); });
```

Comment:
Style nits: inclusion of return in last case could prevent accidental fallthrough if cases are reordered in the future; breaks in place of returns in this case statement seems more conventional to me.